runtime/lua/vim/lsp/sync.lua

```diff
@@ -0,0 +1,363 @@
+  if byte == 1 or #line == 0 then
+    char = byte
+  -- Called in the case of extending an empty line "" -> "a"
+  elseif byte == #line + 1 then
+    byte = byte + vim.str_utf_end(line, #line)
+    char = M.byte_to_utf(line, byte, offset_encoding)
+  else
+    -- Modifying line, find the nearest utf codepoint
+    if align == 'start' then
+      byte = byte + vim.str_utf_start(line, byte)
+      char = M.byte_to_utf(line, byte, offset_encoding)
+    elseif align == 'end' then
+      local offset = vim.str_utf_end(line, byte)
+      -- If the byte does not fall on the start of the character, then
+      -- align to the start of the next character.
+      if offset > 0 then
+        char = M.byte_to_utf(line, byte, offset_encoding) + 1
+        byte = byte + offset
+      else
+        char = M.byte_to_utf(line, byte, offset_encoding)
+        byte = byte + offset
+      end
+    else
+      assert(false, '`align` must be start or end.')
+    end
+    -- Extending line, find the nearest utf codepoint for the last valid character
+  end
+  return byte, char
+end
+
+---@private
+--- Finds the first line, byte, and char index of the difference between the previous and current lines buffer normalized to the previous codepoint.
+---@param prev_lines table list of lines from previous buffer
+---@param curr_lines table list of lines from current buffer
+---@param firstline integer firstline from on_lines, adjusted to 1-index
+---@param lastline integer lastline from on_lines, adjusted to 1-index
+---@param new_lastline integer new_lastline from on_lines, adjusted to 1-index
+---@param offset_encoding string utf-8|utf-16|utf-32|nil (fallback to utf-8)
+---@returns table<int, int> line_idx, byte_idx, and char_idx of first change position
+function M.compute_start_range(prev_lines, curr_lines, firstline, lastline, new_lastline, offset_encoding)
+  -- If firstline == lastline, no existing text is changed. All edit operations
+  -- occur on a new line pointed to by lastline. This occurs during insertion of
+  -- new lines(O), the new newline is inserted at the line indicated by
+  -- new_lastline.
+  -- If firstline == new_lastline, the first change occured on a line that was deleted.
+  -- In this case, the first byte change is also at the first byte of firstline
+  if firstline == new_lastline or firstline == lastline then
+    return { line_idx = firstline, byte_idx = 1, char_idx = 1 }
+  end
+
+  local prev_line = prev_lines[firstline]
+  local curr_line = curr_lines[firstline]
+
+  -- Iterate across previous and current line containing first change
+  -- to find the first different byte.
+  -- Note: *about -> a*about will register the second a as the first
+  -- difference, regardless of edit since we do not receive the first
+  -- column of the edit from on_lines.
+  local start_byte_idx = 1
+  for idx = 1, #prev_line + 1 do
+    start_byte_idx = idx
+    if string.byte(prev_line, idx) ~= string.byte(curr_line, idx) then
```

Comment:
By the way, I am not sure about it, but using `string.byte` from a global namespace may not undergo JIT compilation (because the global has to be queried every time), even though `string.byte` itself is definitely compiled. I would localize it (i.e. put `local string_byte = string.byte` outside of the loop or at the start of the file) just to be safe.